near-contract-standards/src/non_fungible_token/core_impl.rs

```diff
@@ -0,0 +1,442 @@
+    fn nft_on_transfer(
+        &mut self,
+        sender_id: AccountId,
+        previous_owner_id: AccountId,
+        token_id: TokenId,
+        msg: String,
+    ) -> PromiseOrValue<bool>;
+}
+
+/// Implementation of a NonFungibleToken standard.
+/// Allows to include NEP-171 compatible token to any contract.
+/// There are next traits that any contract may implement:
+///     - NonFungibleTokenCore -- interface with nft_transfer methods. NonFungibleToken provides methods for it.
+///     - NonFungibleTokenApproval -- interface with nft_approve methods. NonFungibleToken provides methods for it.
+///     - NonFungibleTokenEnumeration -- interface for getting lists of tokens. NonFungibleToken provides methods for it.
+///     - NonFungibleTokenMetadata -- return metadata for the token in NEP-177, up to contract to implement.
+///
+/// For example usage, see examples/non-fungible-token/src/lib.rs.
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct NonFungibleToken {
+    // owner of contract; this is the only account allowed to call `mint`
+    pub owner_id: AccountId,
+
+    // The storage size in bytes for each new token
+    pub extra_storage_in_bytes_per_token: StorageUsage,
+
+    // always required
+    pub owner_by_id: UnorderedMap<TokenId, AccountId>,
+
+    // required by metadata extension
+    pub token_metadata_by_id: Option<LookupMap<TokenId, TokenMetadata>>,
+
+    // required by enumeration extension
+    pub tokens_per_owner: Option<LookupMap<AccountId, UnorderedSet<TokenId>>>,
+
+    // required by approval extension
+    pub approvals_by_id: Option<LookupMap<TokenId, HashMap<AccountId, u64>>>,
+    pub next_approval_id_by_id: Option<LookupMap<TokenId, u64>>,
+}
+
+#[derive(BorshStorageKey, BorshSerialize)]
+pub enum StorageKeys {
+    TokensForOwner { account_hash: Vec<u8> },
+}
+
+impl NonFungibleToken {
+    pub fn new<Q, R, S, T>(
+        owner_by_id_prefix: Q,
+        owner_id: ValidAccountId,
+        token_metadata_prefix: Option<R>,
+        enumeration_prefix: Option<S>,
+        approval_prefix: Option<T>,
+    ) -> Self
+    where
+        Q: IntoStorageKey,
+        R: IntoStorageKey,
+        S: IntoStorageKey,
+        T: IntoStorageKey,
+    {
+        let (approvals_by_id, next_approval_id_by_id) = if let Some(prefix) = approval_prefix {
+            let prefix: Vec<u8> = prefix.into_storage_key();
+            (
+                Some(LookupMap::new(prefix.clone())),
+                Some(LookupMap::new([prefix, "n".into()].concat())),
+            )
+        } else {
+            (None, None)
+        };
+
+        let mut this = Self {
+            owner_id: owner_id.into(),
+            extra_storage_in_bytes_per_token: 0,
+            owner_by_id: UnorderedMap::new(owner_by_id_prefix),
+            token_metadata_by_id: if let Some(prefix) = token_metadata_prefix {
+                Some(LookupMap::new(prefix))
+            } else {
+                None
+            },
```

Comment:
```suggestion
            token_metadata_by_id: token_metadata_prefix.map(|prefix| Some(LookupMap::new(prefix))),
```

Reply:
Get:

>expected `LookupMap<TokenId, TokenMetadata>`, found `Option<LookupMap<<unknown>, <unknown>>>`

Reply:
```suggestion
            token_metadata_by_id: token_metadata_prefix.map(|prefix| LookupMap::new(prefix)
```
oops, forgot to remove the `Some(..)`, this is it

Reply:
Neat, looks like we were missing a trailing `)` but I wanted to push up a commit anyway since I did this exact pattern right below it after adding the enumeration standard. Thanks!